File: sdks/javascript-admin/src/compliance.ts
/**
 * Compliance service for GDPR, consent management, breach tracking, and auditing.
 */

import type { RequestFn } from './index';

// ============================================================================
// Types
// ============================================================================

// --- Data Subject ---

export interface DataDeletionResult {
  subject_id: string;
  records_deleted: number;
  tables_affected: string[];
  completed_at: string;
}

export interface DataExportRequest {
  subject_id: string;
  format?: string;
  include_metadata?: boolean;
}

export interface DataExportResult {
  subject_id: string;
  format: string;
  data: Record<string, unknown>;
  generated_at: string;
  size_bytes?: number;
}

// --- Certificates ---

export interface ComplianceCertificate {
  id: string;
  type: string;
  standard: string;
  status: string;
  issued_at: string;
  expires_at?: string;
  issuer?: string;
  details?: Record<string, unknown>;
}

export interface CertificateVerification {
  id: string;
  valid: boolean;
  verified_at: string;
  details?: Record<string, unknown>;
}

// --- Audit ---

export interface ComplianceAuditEntry {
  id: string;
  subject_id: string;
  action: string;
  timestamp: string;
  details?: Record<string, unknown>;
  actor?: string;
}

export interface AuditVerification {
  valid: boolean;
  entries_checked: number;
  tampered_entries: number;
  verified_at: string;
}

// --- Consent ---

export interface ConsentRecord {
  subject_id: string;
  purpose: string;
  granted: boolean;
  granted_at?: string;
  revoked_at?: string;
  expires_at?: string;
  source?: string;
  metadata?: Record<string, unknown>;
}

export interface ConsentRequest {
  subject_id: string;
  purpose: string;
  granted: boolean;
  expires_at?: string;
  source?: string;
  metadata?: Record<string, unknown>;
}

export interface ConsentStats {
  total_subjects: number;
  total_consents: number;
  granted: number;
  revoked: number;
  expired: number;
  by_purpose: Record<string, number>;
}

export interface ConsentHistory {
  subject_id: string;
  history: ConsentRecord[];
}

export interface ConsentExport {
  subject_id: string;
  consents: ConsentRecord[];
  exported_at: string;
}

export interface ConsentCheck {
  subject_id: string;
  purpose: string;
  granted: boolean;
  expires_at?: string;
}

// --- Do Not Sell ---

export interface DoNotSellList {
  subjects: DoNotSellEntry[];
  total: number;
}

export interface DoNotSellEntry {
  subject_id: string;
  requested_at: string;
  status: string;
}

// --- Breaches ---

export interface Breach {
  id: string;
  title: string;
  description: string;
  severity: string;
  status: string;
  detected_at: string;
  acknowledged_at?: string;
  resolved_at?: string;
  affected_subjects?: number;
  affected_data_types?: string[];
  reporter?: string;
}

export interface BreachStats {
  total: number;
  open: number;
  acknowledged: number;
  resolved: number;
  by_severity: Record<string, number>;
}

export interface BreachCleanupResult {
  cleaned: number;
  message: string;
}

export interface BreachReport {
  breach: Breach;
  timeline: BreachTimelineEntry[];
  affected_subjects: number;
  notification_status?: string;
  generated_at: string;
}

export interface BreachTimelineEntry {
  timestamp: string;
  action: string;
  actor?: string;
  details?: string;
}

// --- Security Events ---

export interface SecurityEvent {
  id: string;
  timestamp: string;
  event_type: string;
  severity: string;
  source: string;
  description: string;
  details?: Record<string, unknown>;
}

// ============================================================================
// Service
// ============================================================================

export class ComplianceService {
  constructor(private request: RequestFn) {}

  // --- Data Subject Rights ---

  /** Delete all data for a data subject (GDPR right to erasure). */
  async deleteDataSubject(subjectId: string): Promise<DataDeletionResult> {
    return this.request<DataDeletionResult>(
      'DELETE',
      `/api/v1/compliance/data-subject/${encodeURIComponent(subjectId)}`
    );
  }

  /** Export all data for a data subject (GDPR right to portability). */
  async exportData(options: DataExportRequest): Promise<DataExportResult> {
    return this.request<DataExportResult>('POST', '/api/v1/compliance/export', options);
  }

  // --- Certificates ---

  /** List compliance certificates. */
  async listCertificates(): Promise<ComplianceCertificate[]> {
    return this.request<ComplianceCertificate[]>('GET', '/api/v1/compliance/certificates');
  }

  /** Get a specific compliance certificate. */
  async getCertificate(id: string): Promise<ComplianceCertificate> {
    return this.request<ComplianceCertificate>(
      'GET',
      `/api/v1/compliance/certificates/${encodeURIComponent(id)}`
    );
  }

  /** Verify a compliance certificate. */
  async verifyCertificate(id: string): Promise<CertificateVerification> {
    return this.request<CertificateVerification>(
      'GET',
      `/api/v1/compliance/certificates/${encodeURIComponent(id)}/verify`
    );
  }

  // --- Audit ---

  /** Get compliance audit trail for a data subject. */
  async getAuditTrail(subjectId: string): Promise<ComplianceAuditEntry[]> {
    return this.request<ComplianceAuditEntry[]>(
      'GET',
      `/api/v1/compliance/audit/${encodeURIComponent(subjectId)}`
    );
  }

  /** Verify audit log integrity. */
  async verifyAuditLog(): Promise<AuditVerification> {
    return this.request<AuditVerification>('GET', '/api/v1/compliance/audit/verify');
  }

  // --- Consent Management ---

  /** Record or update consent. */
  async recordConsent(consent: ConsentRequest): Promise<ConsentRecord> {
    return this.request<ConsentRecord>('POST', '/api/v1/compliance/consent', consent);
  }

  /** Get consent statistics. */
  async getConsentStats(): Promise<ConsentStats> {
    return this.request<ConsentStats>('GET', '/api/v1/compliance/consent/stats');
  }

  /** Get all consent records for a subject. */
  async getSubjectConsent(subjectId: string): Promise<ConsentRecord[]> {
    return this.request<ConsentRecord[]>(
      'GET',
      `/api/v1/compliance/consent/${encodeURIComponent(subjectId)}`
    );
  }

  /** Delete all consent records for a subject. */
  async deleteSubjectConsent(subjectId: string): Promise<void> {
    await this.request<void>(
      'DELETE',
      `/api/v1/compliance/consent/${encodeURIComponent(subjectId)}`
    );
  }

  /** Get consent history for a subject. */
  async getConsentHistory(subjectId: string): Promise<ConsentHistory> {
    return this.request<ConsentHistory>(
      'GET',
      `/api/v1/compliance/consent/${encodeURIComponent(subjectId)}/history`
    );
  }

  /** Export consent records for a subject. */
  async exportConsent(subjectId: string): Promise<ConsentExport> {
    return this.request<ConsentExport>(
      'GET',
      `/api/v1/compliance/consent/${encodeURIComponent(subjectId)}/export`
    );
  }

  /** Check if a subject has granted consent for a specific purpose. */
  async checkConsent(subjectId: string, purpose: string): Promise<ConsentCheck> {
    return this.request<ConsentCheck>(
      'GET',
      `/api/v1/compliance/consent/${encodeURIComponent(subjectId)}/check/${encodeURIComponent(purpose)}`
    );
  }

  /** Revoke consent for a specific purpose. */
  async revokeConsent(subjectId: string, purpose: string): Promise<void> {
    await this.request<void>(
      'DELETE',
      `/api/v1/compliance/consent/${encodeURIComponent(subjectId)}/${encodeURIComponent(purpose)}`
    );
  }

  // --- Do Not Sell ---

  /** Get the do-not-sell list (CCPA). */
  async getDoNotSellList(): Promise<DoNotSellList> {
    return this.request<DoNotSellList>('GET', '/api/v1/compliance/do-not-sell');
  }

  // --- Breach Management ---

  /** List all breaches. */
  async listBreaches(): Promise<Breach[]> {
    return this.request<Breach[]>('GET', '/api/v1/compliance/breaches');
  }

  /** Get breach statistics. */
  async getBreachStats(): Promise<BreachStats> {
    return this.request<BreachStats>('GET', '/api/v1/compliance/breaches/stats');
  }

  /** Clean up resolved breaches. */
  async cleanupBreaches(): Promise<BreachCleanupResult> {
    return this.request<BreachCleanupResult>('POST', '/api/v1/compliance/breaches/cleanup');
  }

  /** Get a specific breach by ID. */
  async getBreach(id: string): Promise<Breach> {
    return this.request<Breach>(
      'GET',
      `/api/v1/compliance/breaches/${encodeURIComponent(id)}`
    );
  }

  /** Acknowledge a breach. */
  async acknowledgeBreach(id: string): Promise<Breach> {
    return this.request<Breach>(
      'POST',
      `/api/v1/compliance/breaches/${encodeURIComponent(id)}/acknowledge`
    );
  }

  /** Mark a breach as resolved. */
  async resolveBreach(id: string): Promise<Breach> {
    return this.request<Breach>(
      'POST',
      `/api/v1/compliance/breaches/${encodeURIComponent(id)}/resolve`
    );
  }

  /** Generate a formal breach report. */
  async getBreachReport(id: string): Promise<BreachReport> {
    return this.request<BreachReport>(
      'GET',
      `/api/v1/compliance/breaches/${encodeURIComponent(id)}/report`
    );
  }

  // --- Security Events ---

  /** Get security events. */
  async getSecurityEvents(): Promise<SecurityEvent[]> {
    return this.request<SecurityEvent[]>('GET', '/api/v1/compliance/security-events');
  }
}

File: sdks/python-admin/aegis_db_admin/compliance.py
"""
Aegis-DB Admin SDK - Compliance Service

GDPR data subject operations, consent management, breach tracking,
and security event access.

@version 1.0.0
@author AutomataNexus Development Team
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, TYPE_CHECKING

from .types import (
    DeletionCertificate,
    ConsentRecord,
    ConsentStats,
    Breach,
    BreachStats,
    SecurityEvent,
)

if TYPE_CHECKING:
    from .client import AegisAdmin


class ComplianceService:
    """GDPR, consent, breach, and compliance operations."""

    def __init__(self, admin: AegisAdmin) -> None:
        self._admin = admin

    # =========================================================================
    # GDPR - Data Subject
    # =========================================================================

    async def delete_data_subject(self, subject_id: str) -> Dict[str, Any]:
        """Delete all data for a data subject (GDPR right to erasure).

        Args:
            subject_id: The data subject identifier.

        Returns:
            Server response dict (typically includes a deletion certificate id).
        """
        return await self._admin._request(
            "DELETE", f"/api/v1/compliance/data-subject/{subject_id}"
        )

    async def export_data_subject(
        self,
        subject_id: str,
        *,
        format: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Export all data for a data subject (GDPR right to data portability).

        Args:
            subject_id: The data subject identifier.
            format: Export format (e.g. 'json', 'csv').

        Returns:
            Dict containing the exported data or download reference.
        """
        payload: Dict[str, Any] = {"subject_id": subject_id}
        if format is not None:
            payload["format"] = format
        return await self._admin._request(
            "POST", "/api/v1/compliance/export", payload
        )

    # =========================================================================
    # Deletion Certificates
    # =========================================================================

    async def list_deletion_certificates(self) -> List[DeletionCertificate]:
        """List all deletion certificates.

        Returns:
            List of DeletionCertificate objects.
        """
        data = await self._admin._request("GET", "/api/v1/compliance/certificates")
        certs = data if isinstance(data, list) else data.get("certificates", [])
        return [_parse_certificate(c) for c in certs]

    async def get_deletion_certificate(self, cert_id: str) -> DeletionCertificate:
        """Get a specific deletion certificate.

        Args:
            cert_id: The certificate identifier.

        Returns:
            DeletionCertificate object.
        """
        data = await self._admin._request(
            "GET", f"/api/v1/compliance/certificates/{cert_id}"
        )
        return _parse_certificate(data)

    async def verify_deletion_certificate(self, cert_id: str) -> Dict[str, Any]:
        """Verify a deletion certificate.

        Args:
            cert_id: The certificate identifier.

        Returns:
            Verification result dict.
        """
        return await self._admin._request(
            "GET", f"/api/v1/compliance/certificates/{cert_id}/verify"
        )

    # =========================================================================
    # Deletion Audit
    # =========================================================================

    async def get_deletion_audit(self, subject_id: str) -> Dict[str, Any]:
        """Get the deletion audit trail for a data subject.

        Args:
            subject_id: The data subject identifier.

        Returns:
            Audit trail dict.
        """
        return await self._admin._request(
            "GET", f"/api/v1/compliance/audit/{subject_id}"
        )

    async def verify_audit_integrity(self) -> Dict[str, Any]:
        """Verify audit log integrity.

        Returns:
            Verification result dict.
        """
        return await self._admin._request("GET", "/api/v1/compliance/audit/verify")

    # =========================================================================
    # Consent Management
    # =========================================================================

    async def record_consent(
        self,
        subject_id: str,
        purpose: str,
        *,
        granted: bool = True,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Record a consent decision.

        Args:
            subject_id: The data subject identifier.
            purpose: The purpose of consent.
            granted: Whether consent was granted.
            metadata: Optional additional metadata.

        Returns:
            Server response dict.
        """
        payload: Dict[str, Any] = {
            "subject_id": subject_id,
            "purpose": purpose,
            "granted": granted,
        }
        if metadata is not None:
            payload["metadata"] = metadata
        return await self._admin._request(
            "POST", "/api/v1/compliance/consent", payload
        )

    async def get_consent_stats(self) -> ConsentStats:
        """Get consent statistics.

        Returns:
            ConsentStats object.
        """
        data = await self._admin._request("GET", "/api/v1/compliance/consent/stats")
        return ConsentStats(
            total_subjects=data.get("total_subjects", 0),
            total_consents=data.get("total_consents", 0),
            _raw=data,
        )

    async def get_consent_status(self, subject_id: str) -> Dict[str, Any]:
        """Get consent status for a data subject.

        Args:
            subject_id: The data subject identifier.

        Returns:
            Dict with consent status per purpose.
        """
        return await self._admin._request(
            "GET", f"/api/v1/compliance/consent/{subject_id}"
        )

    async def delete_consent_data(self, subject_id: str) -> Dict[str, Any]:
        """Delete all consent records for a data subject.

        Args:
            subject_id: The data subject identifier.

        Returns:
            Server response dict.
        """
        return await self._admin._request(
            "DELETE", f"/api/v1/compliance/consent/{subject_id}"
        )

    async def get_consent_history(self, subject_id: str) -> List[ConsentRecord]:
        """Get consent history for a data subject.

        Args:
            subject_id: The data subject identifier.

        Returns:
            List of ConsentRecord objects.
        """
        data = await self._admin._request(
            "GET", f"/api/v1/compliance/consent/{subject_id}/history"
        )
        records = data if isinstance(data, list) else data.get("history", [])
        return [_parse_consent(r) for r in records]

    async def export_consent(self, subject_id: str) -> Dict[str, Any]:
        """Export consent data for a data subject.

        Args:
            subject_id: The data subject identifier.

        Returns:
            Dict containing exported consent data.
        """
        return await self._admin._request(
            "GET", f"/api/v1/compliance/consent/{subject_id}/export"
        )

    async def check_consent(self, subject_id: str, purpose: str) -> Dict[str, Any]:
        """Check whether a data subject has consented to a specific purpose.

        Args:
            subject_id: The data subject identifier.
            purpose: The purpose to check.

        Returns:
            Dict with consent check result (typically includes 'granted' bool).
        """
        return await self._admin._request(
            "GET", f"/api/v1/compliance/consent/{subject_id}/check/{purpose}"
        )

    async def withdraw_consent(self, subject_id: str, purpose: str) -> Dict[str, Any]:
        """Withdraw consent for a specific purpose.

        Args:
            subject_id: The data subject identifier.
            purpose: The purpose to withdraw consent for.

        Returns:
            Server response dict.
        """
        return await self._admin._request(
            "DELETE", f"/api/v1/compliance/consent/{subject_id}/{purpose}"
        )

    # =========================================================================
    # CCPA
    # =========================================================================

    async def get_do_not_sell_list(self) -> Dict[str, Any]:
        """Get the CCPA do-not-sell list.

        Returns:
            Dict containing the do-not-sell entries.
        """
        return await self._admin._request("GET", "/api/v1/compliance/do-not-sell")

    # =========================================================================
    # Breaches
    # =========================================================================

    async def list_breaches(self) -> List[Breach]:
        """List all breach records.

        Returns:
            List of Breach objects.
        """
        data = await self._admin._request("GET", "/api/v1/compliance/breaches")
        breaches = data if isinstance(data, list) else data.get("breaches", [])
        return [_parse_breach(b) for b in breaches]

    async def get_breach_stats(self) -> BreachStats:
        """Get breach statistics.

        Returns:
            BreachStats object.
        """
        data = await self._admin._request("GET", "/api/v1/compliance/breaches/stats")
        return BreachStats(
            total=data.get("total", 0),
            open=data.get("open", 0),
            acknowledged=data.get("acknowledged", 0),
            resolved=data.get("resolved", 0),
            _raw=data,
        )

    async def cleanup_breaches(self) -> Dict[str, Any]:
        """Run breach cleanup.

        Returns:
            Server response dict.
        """
        return await self._admin._request(
            "POST", "/api/v1/compliance/breaches/cleanup"
        )

    async def get_breach(self, breach_id: str) -> Breach:
        """Get a specific breach record.

        Args:
            breach_id: The breach identifier.

        Returns:
            Breach object.
        """
        data = await self._admin._request(
            "GET", f"/api/v1/compliance/breaches/{breach_id}"
        )
        return _parse_breach(data)

    async def acknowledge_breach(self, breach_id: str) -> Dict[str, Any]:
        """Acknowledge a breach.

        Args:
            breach_id: The breach identifier.

        Returns:
            Server response dict.
        """
        return await self._admin._request(
            "POST", f"/api/v1/compliance/breaches/{breach_id}/acknowledge"
        )

    async def resolve_breach(
        self,
        breach_id: str,
        *,
        resolution: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Resolve a breach.

        Args:
            breach_id: The breach identifier.
            resolution: Optional resolution description.

        Returns:
            Server response dict.
        """
        payload: Dict[str, Any] = {}
        if resolution is not None:
            payload["resolution"] = resolution
        return await self._admin._request(
            "POST", f"/api/v1/compliance/breaches/{breach_id}/resolve", payload
        )

    async def get_breach_report(self, breach_id: str) -> Dict[str, Any]:
        """Get a breach report.

        Args:
            breach_id: The breach identifier.

        Returns:
            Breach report dict.
        """
        return await self._admin._request(
            "GET", f"/api/v1/compliance/breaches/{breach_id}/report"
        )

    # =========================================================================
    # Security Events
    # =========================================================================

    async def get_security_events(self) -> List[SecurityEvent]:
        """Get compliance security events.

        Returns:
            List of SecurityEvent objects.
        """
        data = await self._admin._request(
            "GET", "/api/v1/compliance/security-events"
        )
        events = data if isinstance(data, list) else data.get("events", [])
        return [
            SecurityEvent(
                id=e.get("id"),
                event_type=e.get("event_type"),
                severity=e.get("severity"),
                source_ip=e.get("source_ip"),
                message=e.get("message"),
                timestamp=e.get("timestamp"),
                _raw=e,
            )
            for e in events
        ]


# =============================================================================
# Helpers
# =============================================================================

def _parse_certificate(data: Dict[str, Any]) -> DeletionCertificate:
    return DeletionCertificate(
        id=data.get("id"),
        subject_id=data.get("subject_id"),
        deleted_at=data.get("deleted_at"),
        verified=data.get("verified"),
        _raw=data,
    )


def _parse_consent(data: Dict[str, Any]) -> ConsentRecord:
    return ConsentRecord(
        subject_id=data.get("subject_id"),
        purpose=data.get("purpose"),
        granted=data.get("granted", False),
        timestamp=data.get("timestamp"),
        _raw=data,
    )


def _parse_breach(data: Dict[str, Any]) -> Breach:
    return Breach(
        id=data.get("id"),
        severity=data.get("severity"),
        description=data.get("description"),
        status=data.get("status"),
        detected_at=data.get("detected_at"),
        acknowledged_at=data.get("acknowledged_at"),
        resolved_at=data.get("resolved_at"),
        _raw=data,
    )
